segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -547,366 +678,538 @@ private void validateSystemSnapshot(SystemSnapshotRecord systemSnapshot) {
-                byte[] contents = getContents(systemLogName);
-                val input = new ByteArrayInputStream(contents);
-
-                // Apply record batches from the file.
-                // Loop is exited with eventual EOFException.
-                while (true) {
-                    try {
-                        val batch = BATCH_SERIALIZER.deserialize(input);
-                        if (null != batch.getSystemJournalRecords()) {
-                            for (val record : batch.getSystemJournalRecords()) {
-                                log.debug("SystemJournal[{}] Processing system log record ={}.", epoch, record);
-                                if (visitedRecords.contains(record)) {
-                                    continue;
-                                }
-                                visitedRecords.add(record);
-                                // ChunkAddedRecord.
-                                if (record instanceof ChunkAddedRecord) {
-                                    val chunkAddedRecord = (ChunkAddedRecord) record;
-                                    applyChunkAddition(txn, chunkStartOffsets,
-                                            chunkAddedRecord.getSegmentName(),
-                                            nullToEmpty(chunkAddedRecord.getOldChunkName()),
-                                            chunkAddedRecord.getNewChunkName(),
-                                            chunkAddedRecord.getOffset());
-                                }
-
-                                // TruncationRecord.
-                                if (record instanceof TruncationRecord) {
-                                    val truncationRecord = (TruncationRecord) record;
-                                    finalTruncateOffsets.put(truncationRecord.getSegmentName(), truncationRecord.getOffset());
-                                    finalFirstChunkStartsAtOffsets.put(truncationRecord.getSegmentName(), truncationRecord.getStartOffset());
-                                }
-                            }
-                        }
-                    } catch (EOFException e) {
-                        log.debug("SystemJournal[{}] Done processing file {}.", containerId, systemLogName);
-                        break;
+        val epochToRecover = new AtomicLong(epochToStartScanning.get());
+        return Futures.loop(
+                () -> epochToRecover.get() < epoch,
+                () -> {
+                    // Start scan with file index 1 if epoch is later than snapshot.
+                    if (epochToRecover.get() > epochToStartScanning.get()) {
+                        fileIndexToRecover.set(1);
                     }
-                }
-                // Move to next file.
-                fileIndexToRecover++;
-            }
+
+                    // Process one file at a time.
+                    val isFileDone = new AtomicBoolean();
+                    return Futures.loop(
```

Comment:
Please factor out this big loop into smaller methods. I'm counting 3 nested async For Loops - this is pretty hard to read.

Reply:
done